instance/instance.go

```diff
@@ -92,346 +92,383 @@ func (s *Store) GetList(w http.ResponseWriter, r *http.Request) {
+			if instance.Edition == "" {
+				instance.Edition = currentInstance.Edition
+			}
+			edition := instance.Edition
+			logData["edition"] = edition
+			auditParams["edition"] = edition
+
+			editionDoc, auditAction, err := func() (*models.EditionUpdate, string, error) {
+				// Only create edition if it doesn't already exist
+				editionDoc, action, err := s.getEdition(ctx, datasetID, edition, instanceID)
+				if err != nil {
+					log.ErrorCtx(ctx, errors.WithMessage(err, "instance update: store.getEdition returned an error"), logData)
+					return nil, action, err
+				}
 
-	}
+				// Update with any edition.next changes
+				editionDoc.Next.State = instance.State
+				if err = s.UpsertEdition(datasetID, edition, editionDoc); err != nil {
+					log.ErrorCtx(ctx, errors.WithMessage(err, "instance update: store.UpsertEdition returned an error"), logData)
+					return nil, action, err
+				}
 
-	if instance.State == models.EditionConfirmedState {
-		datasetID := currentInstance.Links.Dataset.ID
+				return editionDoc, action, nil
+			}()
+			if err != nil {
+				if auditAction == "" {
+					// No action to audit so return early
+					return err
+				}
 
-		// If instance has no edition, get the current edition
-		if instance.Edition == "" {
-			instance.Edition = currentInstance.Edition
-		}
-		edition := instance.Edition
+				if auditErr := s.Auditor.Record(ctx, auditAction, audit.Unsuccessful, auditParams); auditErr != nil {
+					err = auditErr
+				}
 
-		// Only create edition if it doesn't already exist
-		editionDoc, err := s.getEdition(ctx, datasetID, edition, id)
-		if err != nil {
-			log.ErrorCtx(ctx, errors.WithMessage(err, "instance update: store.getEdition returned an error"), data)
-			handleInstanceErr(ctx, err, w, data)
-			return
-		}
+				return err
+			}
 
-		// Update with any edition.next changes
-		editionDoc.Next.State = instance.State
-		if err = s.UpsertEdition(datasetID, edition, editionDoc); err != nil {
-			log.ErrorCtx(ctx, errors.WithMessage(err, "instance update: store.UpsertEdition returned an error"), data)
-			handleInstanceErr(ctx, err, w, data)
-			return
-		}
+			s.Auditor.Record(ctx, auditAction, audit.Successful, auditParams)
 
-		data["edition"] = edition
-		log.InfoCtx(ctx, "instance update: created edition", data)
+			log.InfoCtx(ctx, "instance update: created edition", logData)
 
-		// Check whether instance has a version
-		if currentInstance.Version < 1 {
-			// Find the latest version for the dataset edition attached to this
-			// instance and append by 1 to set the version of this instance document
-			version, err := s.GetNextVersion(datasetID, edition)
-			if err != nil {
-				log.ErrorCtx(ctx, errors.WithMessage(err, "instance update: store.GetNextVersion returned an error"), data)
-				handleInstanceErr(ctx, err, w, data)
-				return
+			delete(logData, "dataset_id")
+			delete(logData, "edition")
+			delete(auditParams, "dataset_id")
+			delete(auditParams, "edition")
```

Comment:
Feels a bit weird to be deleting things from the log / audit params. Makes me wonder if we should just go back to defining new instances of data where its needed? I am guessing this is a more specific case though so don't feel the need to change it - was more just an thought for ongoing work / maintanence

Reply:
I can create a second log and audit params for inline function and hence not have to delete parameters